scripts/gen.py

```diff
@@ -0,0 +1,536 @@
+  for p in params:
+    ptype = p.children[0]
+    cptype = type_core(ptype)
+    pname = param_name(p)
+    if cptype == 'TensorOptions' or cptype == 'TensorList':
+      other = pname
+    if cptype != 'Tensor':
+      continue
+    if pname == 'self':
+      return pname
+    if type_is_const(ptype):
+      ref_param = pname
+  return ref_param or other
+
+
+def get_tuple_return(rtype, rtype_str, rname, params, param_vars, ref_param):
+  types = tuple_type_list(rtype)
+  retstr = '{}('.format(rtype_str)
+  for i, ttype in enumerate(types):
+    if i > 0:
+      retstr += ', '
+    tuple_var = 'std::get<{}>({})'.format(i, rname)
+    retstr += get_return_value(ttype, tuple_var, list_get(params, i),
+                               list_get(param_vars, i), ref_param)
+  return retstr + ')'
+
+
+def get_return_type_str(t, orig_sig):
+  assert isinstance(t, lark.tree.Tree)
+  fname = t.children[1]
+  assert isinstance(fname, lark.tree.Tree)
+  assert fname.data == 'fnname'
+  token = fname.children[0]
+  assert isinstance(token, lark.lexer.Token)
+  return orig_sig[0:token.column - 2]
+
+
+def generate_return_stmt(t, orig_sig, fname, rname, params, param_vars,
+                         ref_param):
+  assert isinstance(t, lark.tree.Tree)
+  rtype = t.children[0]
+  ctype = type_core(rtype)
+  if ctype == 'std::tuple':
+    rtype_str = get_return_type_str(t, orig_sig)
+    retstr = get_tuple_return(rtype, rtype_str, rname, params, param_vars,
+                              ref_param)
+  elif ctype == 'std::vector':
+    retstr = 'bridge::CreateXlaTensors({}, bridge::XlaTensorDevice({}))'.format(
+        rname, ref_param)
+  elif ctype == 'Tensor':
+    retstr = get_return_value(rtype, rname, params[0], param_vars[0], ref_param)
+  elif ctype == 'void' and not type_is_refptr(rtype, '*'):
+    return ''
+  else:
+    retstr = rname
+  return '  return {};\n'.format(retstr)
+
+
+def generate_result_assignment(t, rname):
+  assert isinstance(t, lark.tree.Tree)
+  rtype = t.children[0]
+  ctype = type_core(rtype)
+  if ctype == 'void' and not type_is_refptr(rtype, '*'):
+    return ''
+  return 'auto&& {} = '.format(rname)
+
+
+def get_xla_wrapper(orig_sig, ctx):
+  tree = _PARSER.parse(orig_sig)
+  xtree = _XPARSER.parse(orig_sig)
+  params = get_parameters(tree)
+  ref_param = get_reference_param(params)
+
+  # There are a few functions with the same function name but different
+  # parameter list. Generate a unique XL function name here.
+  def gen_fnname(x):
+    post = ''
+    if x in ctx.defdb:
+      post = '_{}'.format(ctx.defdb[x])
+      ctx.defdb[x] += 1
+    else:
+      ctx.defdb[x] = 1
+    return 'xla_' + x + post
+
+  sig, fname, xfname = get_function_signature(tree, orig_sig, gen_fnname)
+  code = '{} {{\n'.format(sig)
+  xla_ref_param = ref_param
+  param_vars = []
+  for p in params:
+    ptype = p.children[0]
+    cptype = type_core(ptype)
+    pname = param_name(p)
+    if cptype == 'TensorList':
+      xname = '_l_{}'.format(pname)
```

Comment:
Could we avoid prefixing with underscores? It's not allowed by the C++ standard. Same for double underscores anywhere in the identifier, would be nice if we avoided them.

Reply:
SG